aten/src/ATen/CPUGenerator.cpp

```diff
@@ -1,49 +1,87 @@
 #include <ATen/CPUGenerator.h>
-
-#define const_generator_cast(generator) \
-  dynamic_cast<const CPUGenerator&>(generator)
+#include <c10/util/C++17.h>
 
 namespace at {
 
-CPUGenerator::CPUGenerator(Context * context_)
-  : context(context_), generator(THGenerator_new())
-{}
+namespace detail {
 
-CPUGenerator::~CPUGenerator() {
-  if (generator)
-    THGenerator_free(generator);
-}
+// Ensures default_gen_cpu is initialized once.
+static std::once_flag cpu_device_flag;
+
+// Default, global CPU generator.
+static std::unique_ptr<CPUGenerator> default_gen_cpu;
 
-CPUGenerator& CPUGenerator::copy(const Generator& from) {
-  THGenerator_copy(generator, const_generator_cast(from).generator);
-  return *this;
+/**
+ * PyTorch maintains a collection of default generators that get
+ * initialized once. The purpose of these default generators is to
+ * maintain a global running state of the pseudo random number generation,
+ * when a user does not explicitly mention any generator.
+ * getDefaultCPUGenerator gets the default generator for a particular
+ * device.
+ */
+std::unique_ptr<CPUGenerator>& getDefaultCPUGenerator() {
+  std::call_once(cpu_device_flag, [&] {
+    default_gen_cpu = c10::guts::make_unique<CPUGenerator>(default_rng_seed_val);
+  });
+  return default_gen_cpu;
 }
 
-CPUGenerator& CPUGenerator::free() {
-  THGenerator_free(generator);
-  return *this;
+/**
+ * Utility to create a CPUGenerator. Returns a unique_ptr
+ */
+std::unique_ptr<CPUGenerator> createCPUGenerator(uint64_t seed_val) {
+  return c10::guts::make_unique<CPUGenerator>(seed_val, Philox4_32_10(seed_val));
 }
 
-uint64_t CPUGenerator::seed() {
-  return THRandom_seed(generator);
+} // namespace detail
+
+/**
+ * CPUGenerator class implementation
+ */
+CPUGenerator::CPUGenerator(uint64_t seed_in, Philox4_32_10 engine_in)
+  : CloneableGenerator(Device(DeviceType::CPU)), current_seed_(seed_in), engine_(engine_in) {}
+
+/**
+ * Manually seeds the engine with the seed input
+ */
+void CPUGenerator::setCurrentSeed(uint64_t seed) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  current_seed_ = seed;
+  engine_ = Philox4_32_10(seed);
 }
 
-uint64_t CPUGenerator::initialSeed() {
-  return THRandom_initialSeed(generator);
+/*
+ * Gets the current seed of CPUGenerator.
+ */
+uint64_t CPUGenerator::getCurrentSeed() const {
+  std::lock_guard<std::mutex> lock(mutex_);
+  return current_seed_;
 }
 
-CPUGenerator& CPUGenerator::manualSeed(uint64_t seed) {
-  THRandom_manualSeed(generator, seed);
-  return *this;
+/* 
+* Gets a random 32 bit unsigned integer from the engine
+*/
+uint32_t CPUGenerator::random() {
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
The new locking strategy is seriously problematic. You are taking out a lock for *every single* random number you request from the RNG. Yes locking/unlocking uncontended mutexes is cheap, but it's not "you can run it in a tight loop when you need to generate random numbers cheap". I'd recommend doing some benchmarks, e.g., with bernoulli (which I see has been changed from taking out a lock once at the beginning of the kernel, to taking out the lock for every random number to be generated) to see the effect of this.

Reply:
yes! I was about to bring this up but wanted to wait to get through the minor fixes first. But let's start this conversation :). I did benchmark and see a ~1 order magnitude slowdown as expected (normal(), random() etc.) :'(. I want the users to not do 
```
std::lock_guard<std::mutex> lock(generator->mutex);
```
before every call to random()/random64() (as I know that will forget to add this line in my code XD). Do you know of a way that would avoid calling lock_guard before every random call or are we stuck with explicitly putting this lock_guard everywhere before the start of a tight loop (And expect users to read a Note about it)?

Reply:
Syed and I chatted about this in person, and the conclusion was that we should preserve the existing locking behavior for now, and then in a later patch set look into changing the threading semantics of generator (e.g., making them non-thread safe and instead making the generator state splittable, to accommodate forks into other threads.)